larq_compute_engine/mlir/tf_to_tfl_flatbuffer.cc

```diff
@@ -70,25 +78,64 @@ Status ConvertTFExecutorToFlatbuffer(mlir::ModuleOp module, bool export_to_mlir,
 
   mlir::StatusScopedDiagnosticHandler statusHandler(module.getContext(),
                                                     /*propagate=*/true);
+  if (failed(IsValidGraph(module))) {
+    return statusHandler.ConsumeStatus();
+  }
+
+  mlir::PassManager pass_manager(module.getContext());
+  mlir::applyPassManagerCLOptions(pass_manager);
+  pass_manager.addInstrumentation(
+      std::make_unique<mlir::TFL::ErrorCollectorInstrumentation>(
+          pass_manager.getContext()));
 
-  if (failed(IsValidGraph(module)) || failed(pass_manager->run(module))) {
+  tensorflow::AddPreVariableFreezingTFToLCETFLConversionPasses(&pass_manager);
+  if (failed(pass_manager.run(module))) {
     return statusHandler.ConsumeStatus();
   }
 
+  // Freeze variables if a session is provided.
+  if (session.hasValue()) {
+    mlir::TFL::ErrorCollectorInstrumentation collector(module.getContext());
+    if (failed(mlir::tf_saved_model::FreezeVariables(module,
+                                                     session.getValue()))) {
+      auto status = statusHandler.ConsumeStatus();
+      mlir::TFL::ErrorCollector* collector =
+          mlir::TFL::ErrorCollector::GetErrorCollector();
+      if (!collector->CollectedErrors().empty()) {
+        return errors::InvalidArgument("Variable constant folding is failed.");
+      }
+      return status;
+    }
+  }
+  pass_manager.clear();
+  tensorflow::AddPostVariableFreezingTFToLCETFLConversionPasses(
+      saved_model_dir, quant_specs, &pass_manager, target);
+  if (failed(pass_manager.run(module))) {
+    auto status = statusHandler.ConsumeStatus();
+    mlir::TFL::ErrorCollector* collector =
+        mlir::TFL::ErrorCollector::GetErrorCollector();
+    for (const auto& error_data : collector->CollectedErrors()) {
```

Comment:
Is it intentional that this error collection stuff is done if the second phase failed, but not if the first phase failed (i.e. line 93)?

Reply:
To be fair, I don't know. This is primarily [copied from TensorFlow](https://github.com/tensorflow/tensorflow/blob/3f878cff5b698b82eea85db2b60d65a2e320850e/tensorflow/compiler/mlir/lite/tf_to_tfl_flatbuffer.cc#L247-L251) and they seem to do it like this.